### ISSUE TITLE
Subquery improvement for HqlExpressions

### ISSUE THREAD

**Issue by GiuseppeMusso-Laser** [OrchardCMS/Orchard#8162]
Hi all,
I need to filter content items according to a custom HQL query that can extract (potentially) many ids.
So I created a query filter provider that executes my HQL query, gets a list of content ids and adds a where condition in the form of:

  predicate = x => x.InG("Id", ids);

where ids is a variable of type List<int>.
If ids is a very long list, this query may experience some performance problem or get into an error.
So, what if I add a method "InSubquery(string propertyName, string subquery, Dictionary<string, object> parameters)" to IHqlExpressionFactory and implement it in DefaultHqlQuery?

I'll post a PR soon.

**Comment by sebastienros** [comment]
I don't understand how the query is rewritten. Can you share an example?

**Comment by GiuseppeMusso-Laser** [comment]
Hi, here is an example.
We wrote an Orchard Module to deal with user contacts. Every contact (MobileContactPart) may have zero, one or more mobile devices (PushNotificationRecord) associated with it.
I need a projection to extract all contacts with a specific device type (Android, iOS, WindowsMobile) so I wrote an IFilterProvider:

```
public void ApplyFilter(dynamic context) {
    var query = (IHqlQuery)context.Query;

    string subquery = @"SELECT contact.MobileContactPartRecord_Id as contactId
                        FROM Laser.Orchard.Mobile.Models.PushNotificationRecord AS contact 
                        WHERE Device = :device";

    Dictionary<string, object> parameters = new Dictionary<string, object>();
    parameters.Add("device", Convert.ToString(context.State.DeviceType));

    context.Query = query.Where(x => x.ContentPartRecord<CommunicationContactPartRecord>(), x => x.InSubquery("Id", subquery, parameters));
}
```

where the :device parameter is the device type (Android, Apple or WindowsMobile).
In a couple of tenants, we have some thousands of registered devices so I'd rather use a subquery than an array of ids. 

This is just a simplified example, of course.
I could use a pretty complex HQL query to extract content ids according to some where conditions in AND/OR and I could use the full syntax of HQL expressions as well.

**Comment by BenedekFarkas** [comment]
I made a mistake with the milestones - this was actually released in 1.10.3 (PR was merged two weeks before that), so I'll reassign the appropriate milestone and close it.
